src/Core/Persistence/JsonFileHandler.cs

```diff
@@ -0,0 +1,107 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.ComponentModel.Composition;
+using System.IO;
+using System.IO.Abstractions;
+using Newtonsoft.Json;
+
+namespace SonarLint.VisualStudio.Core.Persistence;
+
+public interface IJsonFileHandler
+{
+    bool TryReadFile<T>(string filePath, out T content) where T : class;
+    bool TryWriteToFile<T>(string filePath, T model) where T : class;
+}
+
+[Export(typeof(IJsonFileHandler))]
+[PartCreationPolicy(CreationPolicy.Shared)]
+public class JsonFileHandler : IJsonFileHandler
+{
+    private readonly ILogger logger;
+    private readonly IFileSystem fileSystem;
+    private readonly IJsonSerializer jsonSerializer;
+    private static readonly object Locker = new();
+
+    [ImportingConstructor]
+    public JsonFileHandler(IJsonSerializer jsonSerializer, ILogger logger) : this(new FileSystem(), jsonSerializer, logger) { }
+
+    internal /* for testing */ JsonFileHandler(IFileSystem fileSystem, IJsonSerializer jsonSerializer, ILogger logger)
+    {
+        this.fileSystem = fileSystem;
+        this.jsonSerializer = jsonSerializer;
+        this.logger = logger;
+    }
+
+    public bool TryReadFile<T>(string filePath, out T content) where T: class
+    {
+        content = null;
+        if (!fileSystem.File.Exists(filePath))
```

Comment:
it might be more performant to just try to read regardless of if the file exists or not and handle the exception in the odd case that the file does not exist, because generally the file should exist and you will have an unnecessary file system call here

Reply:
I understand your point, but I don't think it's worth changing it for the following reasons:

- This generic handler might be used in more places, so even if we expect the connections.json to exist, it might not be the expected behavior for other cases
- Exception handling is an expensive operation as well, so it might be that the if check will be faster than handling exceptions
- Generally relying on exceptions for the normal flow of control is not recommended: https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/exception-throwing

Reply:
1) sure
2) exception handling is not as expensive as file system lookup, false equivalence
3) it's about design of public APIs. The Json serialization method should not rethrow it, and it would not, but it would be fine to handle by using the exception IF 1. exceptions would be rare 2. the alternative is a slow file system lookup

Reply:
Do you have a reference that proves that "exception handling is not as expensive as file system lookup"? I think we are squeezing performance here, which is not really needed.

It is not about designing of public APIs. It's a general guideline.

Reply:
As the 1st assumption about rarity of misses is wrong, which you pointed out correctly, you can disregard this thread

Reply:
> Do you have a reference that proves that "exception handling is not as expensive as file system lookup"? I think we are squeezing performance here, which is not really needed.
> 
> It is not about designing of public APIs. It's a general guideline.

Exception handling is just some CPU-bound work, while file system look up calls win api